src/main/scala/is/hail/variant/HTSGenotypeView.scala

```diff
@@ -276,4 +278,115 @@ class ArrayGenotypeView(rowType: TStruct) {
     val elementOffset = ArrayGenotypeView.tArrayFloat64.elementOffset(adOffset, length, idx)
     m.loadDouble(elementOffset)
   }
+}
+
+object HardCallView {
+  def apply(rowSignature: TStruct): HardCallView = {
+    rowSignature.fields(3).typ.asInstanceOf[TArray].elementType match {
+      case TGenotype => new HardCallTGenotypeView(rowSignature)
+      case _: TStruct => new HardCallStructView(rowSignature, "GT")
+      case t => fatal(s"invalid genotype representation: $t, expect TGenotype or TStruct")
+    }
+  }
+}
+
+abstract class HardCallView {
+  def setRegion(mb: MemoryBuffer, offset: Long)
+
+  def setRegion(rv: RegionValue): Unit = setRegion(rv.region, rv.offset)
+
+  def setGenotype(idx: Int)
+
+  def hasGT: Boolean
+
+  def getGT: Int
+}
+
+class HardCallTGenotypeView(rowType: TStruct) extends HardCallView {
+  private val tgs = rowType.fields(3).typ.asInstanceOf[TArray]
```

Comment:
Why `fields(3).typ` here and `fieldType(3)` below in `HardCallStructView`?

Reply:
The row type we use is a struct with four fields:
pk (partition key)
v (row key, variant usually)
va (annotations)
gs (TArray(genotype type))

I want the genotype type, so I need to index in and cast and get out the element type.

Reply:
Sorry, this is such a minor thing, I just want to be sure I'm not missing something important. `rowType.fieldType(3)` is definitionally the same as `rowType.fields(3).typ`, no?

Reply:
Oh, I totally misunderstood your question being about the mismatch between fields(3).typ and fieldType(3). They are the same thing, I didn't realize we had the helper method fieldType, and copied some of the code for HardCallStructView from ArrayGenotypeStructView. 

I'll change the `fields(3).typ`.